File: ravascript/tests/stuff/structs_and_impl_methods/structs_and_impl_methods.rs
use ravascript::prelude::web::Console;

trait MyTrait {
    fn get_age(&self) -> i32;
}
struct MyStruct {
    age: i32,
    name: &'static str,
}
impl MyStruct {
    fn new(age: i32, name: &'static str) -> MyStruct {
        MyStruct { age, name }
    }
    fn my_method(&self) -> &str {
        self.name
    }
    fn my_method_with_arg(&self, inc: i32) -> i32 {
        self.age + inc
    }
    fn my_associated_method(inc: i32) -> i32 {
        inc + 10
    }
    fn with_generic<T>(&self, inc: T) -> i32 {
        self.age
    }
}
// TODO currently impls for a struct must appear directly after it. defining a trait inbetween like below is not suppported
// trait MyTrait {
//     fn get_age(&self) -> i32;
// }
impl MyTrait for MyStruct {
    fn get_age(&self) -> i32 {
        self.age
    }
}

fn main() {
    let thing = MyStruct::new(2, "Bruce");
    // Console::assert(thing.my_method() == "Bruce");
    // Console::assert(thing.my_method_with_arg(2) == 4);
    // Console::assert(MyStruct::my_associated_method(2) == 12);
    // Console::assert(thing.with_generic(99) == 2);
    // Console::assert(thing.get_age() == 2);

    // console.assert(thing.myMethod() === "Bruce");
    // console.assert(thing.myMethodWithArg(2) === 4);
    // console.assert(MyStruct.myAssociatedMethod(2) === 12);
    // console.assert(thing.withGeneric(99) === 2);
    // console.assert(thing.getAge() === 2);

    assert_eq!(thing.my_method(), "Bruce");
    assert_eq!(thing.my_method_with_arg(2), 4);
    assert_eq!(MyStruct::my_associated_method(2), 12);
    assert_eq!(thing.with_generic(99), 2);
    assert_eq!(thing.get_age(), 2);
}

File: ravascript/tests/stuff/structs_and_impl_methods/structs_and_impl_methods.js
class RustInteger {
    constructor(jsNumber) {
        this.jsNumber = jsNumber;
    }

    eq(other) {
        return new RustBool(this.jsNumber === other.jsNumber);
    }
    ne(other) {
        return new RustBool(this.jsNumber !== other.jsNumber);
    }
    add(other) {
        return new RustInteger(this.jsNumber + other.jsNumber);
    }
    addAssign(other) {
        this.jsNumber += other.jsNumber;
    }
    abs() {
        return new RustInteger(Math.abs(this.jsNumber));
    }
    rem(other) {
        return new RustInteger(this.jsNumber % other.jsNumber);
    }
}
class RustString {
    constructor(jsString) {
        this.jsString = jsString;
    }

    eq(other) {
        return new RustBool(this.jsString === other.jsString);
    }
    ne(other) {
        return new RustBool(this.jsString !== other.jsString);
    }
    toString() {
        return this.clone();
    }
    clone() {
        return new RustString(this.jsString);
    }
}
class RustBool {
    constructor(jsBoolean) {
        this.jsBoolean = jsBoolean;
    }

    eq(other) {
        return new RustBool(this.jsBoolean.eq(other.jsBoolean));
    }
    ne(other) {
        return new RustBool(this.jsBoolean.ne(other.jsBoolean));
    }
    boolAnd(other) {
        return new RustBool(this.jsBoolean && other.jsBoolean);
    }
}

class MyStruct {
    constructor(age, name) {
        this.age = age;
        this.name = name;
    }

    static new(age, name) {
        return new MyStruct(age, name);
    }
    myMethod() {
        return this.name;
    }
    myMethodWithArg(inc) {
        return this.age.add(inc);
    }
    static myAssociatedMethod(inc) {
        return inc.add(new RustInteger(10));
    }
    withGeneric(inc) {
        return this.age;
    }
    getAge() {
        return this.age;
    }
}
(function main() {
    var thing = MyStruct.new(new RustInteger(2), new RustString("Bruce"));
    console.assert(thing.myMethod().eq(new RustString("Bruce")));
    console.assert(thing.myMethodWithArg(new RustInteger(2)).eq(new RustInteger(4)));
    console.assert(MyStruct.myAssociatedMethod(new RustInteger(2)).eq(new RustInteger(12)));
    console.assert(thing.withGeneric(new RustInteger(99)).eq(new RustInteger(2)));
    console.assert(thing.getAge().eq(new RustInteger(2)));
})();
